gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcReadChannel.java

```diff
@@ -536,48 +565,46 @@ private boolean isByteBufferBeyondCurrentRequestRange(ByteBuffer byteBuffer) {
-        if (bytesToSkipBeforeReading >= 0 && bytesToSkipBeforeReading < content.size()) {
-          content = res.getChecksummedData().getContent().substring((int) bytesToSkipBeforeReading);
-          positionInGrpcStream += bytesToSkipBeforeReading;
-          bytesToSkipBeforeReading = 0;
-        } else if (bytesToSkipBeforeReading >= content.size()) {
-          positionInGrpcStream += content.size();
-          bytesToSkipBeforeReading -= content.size();
-          continue;
-        }
+    // When zero-copy marshaller is used, the stream that backs GetObjectMediaResponse
+    // should be closed when the message is no longed needed so that all buffers in the
+    // stream can be reclaimed. If zero-copy is not used, stream will be null.
+    InputStream stream = getObjectMediaResponseMarshaller.popStream(res);
+    try {
+      ByteString content = res.getChecksummedData().getContent();
+      if (bytesToSkipBeforeReading >= 0 && bytesToSkipBeforeReading < content.size()) {
+        content = res.getChecksummedData().getContent().substring((int) bytesToSkipBeforeReading);
+        positionInGrpcStream += bytesToSkipBeforeReading;
+        bytesToSkipBeforeReading = 0;
+      } else if (bytesToSkipBeforeReading >= content.size()) {
+        positionInGrpcStream += content.size();
+        bytesToSkipBeforeReading -= content.size();
+        return bytesRead;
+      }
 
-        if (readOptions.isGrpcChecksumsEnabled() && res.getChecksummedData().hasCrc32C()) {
-          validateChecksum(res);
-        }
+      if (readOptions.isGrpcChecksumsEnabled() && res.getChecksummedData().hasCrc32C()) {
+        validateChecksum(res);
+      }
 
-        boolean responseSizeLargerThanRemainingBuffer = content.size() > byteBuffer.remaining();
-        int bytesToWrite =
-            responseSizeLargerThanRemainingBuffer ? byteBuffer.remaining() : content.size();
-        put(content, 0, bytesToWrite, byteBuffer);
-        bytesRead += bytesToWrite;
-        positionInGrpcStream += bytesToWrite;
+      boolean responseSizeLargerThanRemainingBuffer = content.size() > byteBuffer.remaining();
+      int bytesToWrite =
+          responseSizeLargerThanRemainingBuffer ? byteBuffer.remaining() : content.size();
+      put(content, 0, bytesToWrite, byteBuffer);
+      bytesRead += bytesToWrite;
+      positionInGrpcStream += bytesToWrite;
 
-        if (responseSizeLargerThanRemainingBuffer) {
-          invalidateBufferedContent();
-          bufferedContent = content;
-          bufferedContentReadOffset = bytesToWrite;
-          // This is to keep the stream alive for the message backed by this.
-          streamForBufferedContent = stream;
-          stream = null;
-        }
-      } finally {
-        if (stream != null) {
-          stream.close();
-        }
+      if (responseSizeLargerThanRemainingBuffer) {
+        invalidateBufferedContent();
+        bufferedContent = content;
+        bufferedContentReadOffset = bytesToWrite;
```

Comment:
Just like we slice off the content when `bytesToSkipBeforeReading` is > 0, wouldn't it be consistent to do the same thing here and keep bufferedContentReadOffset 0.
```
  bufferredContent = content.substring(bytesToWrite)
```

Reply:
In case of exceptions later on, where we would have to re-populate the buffer, we would lose the data. Holding on to the gcs data one chunk at a time should not be that expensive